dali/test/python/test_pipeline_rnnt.py

```diff
@@ -0,0 +1,161 @@
+        self.splicing_reshape = ops.Reshape(device="cpu", rel_shape=[-1, frame_splicing_factor])
+        self.splicing_pad = ops.Pad(axes=[0], fill_value=0, align=frame_splicing_factor, shape=[1],
+                                    device="cpu")
+
+        self.get_nonsilent_region = ops.NonsilentRegion(device="cpu", cutoff_db=silence_threshold)
+        self.trim_silence = ops.Slice(device="cpu", normalized_anchor=False, normalized_shape=False,
+                                      axes=[0], image_type=types.ANY_DATA)
+        self.to_float = ops.Cast(dtype=types.FLOAT)
+
+    @staticmethod
+    def _div_ceil(dividend, divisor):
+        return (dividend + divisor - 1) // divisor
+
+    def _splice_frames(self, input):
+        """
+        Frame splicing is implemented by transposing the input, padding it and reshaping.
+        Theoretically, to achieve the result now there should be one more transpose at the end,
+        but it can be skipped as an optimization
+        """
+        out = self.splicing_transpose(input)
+
+        # Because of the padding, we need to determine length of audio sample before it occurs
+        audio_len = self._div_ceil(self.get_shape(out), self.frame_splicing_factor)
+
+        out = self.splicing_pad(out)
+        out = self.splicing_reshape(out)
+        # Skipping transposing back
+        return out, audio_len
+
+    def remove_silence(self, input):
+        begin, len = self.get_nonsilent_region(input)
+        out = self.trim_silence(input, self.to_float(begin), self.to_float(len))
+        return out
+
+    def define_graph(self):
+        input, label = self.read()
+        # No resampling (aka "speed perturbation"), because of randomness of this operation
+        audio, sr = self.decode(input)
+
+        audio = self.remove_silence(audio)
+
+        audio = audio + self.normal_distribution(audio) * self.dither
+
+        # DALI's preemph works a little bit different than the one in native code.
+        # The difference occurs in first value in buffer.
+        audio = self.preemph(audio)
+
+        audio = self.spectrogram(audio)
+        audio = self.mel_fbank(audio)
+        audio = self.log_features(audio)
+        audio, audio_sh = self._splice_frames(audio)
+
+        # This normalization goes across ax=0, since
+        # the frame splicing doesn't transpose the tensor back
+        audio = self.normalize(audio)
+
+        return audio.gpu(), audio_sh.gpu()
+
+
+def test_rnnt_data_pipeline():
+    """
+    Test compares pre-calculated output of native data pipeline with an output
+    from DALI data pipeline. There are few modification of native data pipeline
+    comparing to the reference:
+    1. Presampling (aka "speed perturbation") is turned off
+    2. Since DALI, as an optimization, doesn't perform one transposition in frame splicing,
+       the result from DALI pipeline has to be transposed to fit the reference data.
```

Comment:
Why not to transpose back in this test pipeline? Do we have it super optimized here?

Reply:
That's rather a concept-optimization. Reference has 2 additional transpositions, since there's a problem, which doesn't occurr in DALI: equal sample shape in batch constrain. It's better this way, prooves that getting rid of double transpose is feasible

Reply:
But still we have one transpose in the test validation. So why not to move it to the pipeline?

Reply:
Outdated in new code